bottlerocket-settings-models/settings-extensions/bootstrap-commands/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+    use serde_json::json;
+
+    #[test]
+    fn test_generate_bootstrap_command_settings() {
+        let generated = BootstrapCommandsSettingsV1::generate(None, None).unwrap();
+
+        assert_eq!(
+            generated,
+            GenerateResult::Complete(BootstrapCommandsSettingsV1 {
+                bootstrap_commands: BTreeMap::new(),
+            })
+        )
+    }
+
+    #[test]
+    fn test_serde_bootstrap_command() {
+        let test_json = json!({
+            "mybootstrap": {
+                "commands": [ ["apiclient", "motd=hello"], ],
+                "mode": "once",
+                "essential": true,
+            }
+        });
+
+        let bootstrap_commands: BootstrapCommandsSettingsV1 =
+            serde_json::from_value(test_json.clone()).unwrap();
+
+        let mut expected_bootstrap_commands: BTreeMap<Identifier, BootstrapCommand> =
+            BTreeMap::new();
+        expected_bootstrap_commands.insert(
+            Identifier::try_from("mybootstrap").unwrap(),
+            BootstrapCommand {
+                commands: Some(vec![ValidCommand::try_from(vec![
+                    "apiclient".to_string(),
+                    "motd=hello".to_string(),
+                ])
+                .unwrap()]),
+                mode: Some(BootstrapMode::try_from("once").unwrap()),
+                essential: Some(true),
+            },
+        );
+
+        assert_eq!(
+            bootstrap_commands,
+            BootstrapCommandsSettingsV1 {
+                bootstrap_commands: expected_bootstrap_commands
+            }
+        );
+
+        let serialized_json: serde_json::Value = serde_json::to_string(&bootstrap_commands)
+            .map(|s| serde_json::from_str(&s).unwrap())
+            .unwrap();
+
+        assert_eq!(serialized_json, test_json);
+    }
+
+    #[test]
+    fn test_serde_invalid_bootstrap_command() {
+        let test_err_json = json!({
+            "mybootstrap1": {
+                "commands": [ ["/usr/bin/touch", "helloworld"], ],
+                "mode": "once",
+                "essential": true,
+            }
+        });
+
+        let bootstrap_commands_err: std::result::Result<
+            BootstrapCommandsSettingsV1,
+            serde_json::Error,
+        > = serde_json::from_value(test_err_json.clone());
+
+        // This has invalid command. It should fail.
+        assert!(bootstrap_commands_err.is_err());
+    }
+}
+
+/// ValidCommand represents a valid Bootstrap Command. It stores the command as a vector of
+/// strings and ensures that the first argument is apiclient.
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Default, Serialize)]
+pub struct ValidCommand(Vec<String>);
+
+impl ValidCommand {
+    pub fn get_args(&self) -> (String, Vec<String>) {
+        self.0
+            .clone()
+            .split_first()
+            .map_or(("".to_string(), Vec::new()), |(command, arguments)| {
+                (command.to_string(), arguments.to_vec())
+            })
+    }
+}
```

Comment:
I have a few nits with this function:
1. The name is a bit misleading, since it really separates the command from the args
2. It unnecessarily clones data (both in the implementation as well as the interface).


For (1), you could implement this instead as two different functions `get_command()` and `get_args()`, or else I might rename this as `get_command_and_args()`.

For (2), the clone you have is unnecessary:

```suggestion
    pub fn get_args(&self) -> (String, Vec<String>) {
        self.0
            .split_first()
            .map_or(("".to_string(), Vec::new()), |(command, arguments)| {
                (command.to_string(), arguments.to_vec())
            })
    }
}
```

but I might recommend instead changing the interface to this:

```suggestion
    pub fn get_command_and_args(&self) -> (&str, &[String]) {
        self.0
            .split_first()
            .map(|(command, rest)| (command.as_str(), rest))
            .unwrap_or_else(Default::default)
    }
}
```

If possible, it's best to let the caller decide to allocate.

Reply:
I like your last implementation. `pub fn get_command_and_args(&self) -> (&str, &[String])`. It will help clean this up as well -> https://github.com/piyush-jena/bottlerocket-core-kit/blob/a4a72e2f1d6336c2669b4419bb1485f95cdfb824/sources/api/bootstrap-commands/src/main.rs#L113. args was a force of habit from C++ where argv contained the command as well as its first element.

Reply:
For the linked code I would expect `String` to work without having to call `as_str` on everything, since `String` [has an impl for `AsRef<OsStr>`](https://doc.rust-lang.org/std/string/struct.String.html#impl-AsRef%3COsStr%3E-for-String).

Regardless, the `String` slice should work as well. Just call `iter()` on the result.

Reply:
Updated with the last suggestion with a small change. `unwrap_or_default()` instead of `unwrap_or_else(Default::default)`